src/ship/GunManager.h

```diff
@@ -0,0 +1,222 @@
+
+		//TODO: integrate this with a whole-ship cooling system
+		float coolingPerSecond = 14.2; // nominal amount of thermal energy removed per second (kW)
+		float overheatThreshold = 280; // total amount of thermal energy(kJ) the gun can store while functioning
+
+		std::string modelPath; // model to render this weapon with
+
+		ProjectileData projectile; // deprecated, to replace with RefCountedPtr<ProjectileDef>
+		ProjectileType projectileType = PROJECTILE_BALLISTIC;
+		uint8_t numBarrels = 1;		 // total number of barrels on the model
+		bool staggerBarrels = false; // should we fire one barrel after another, or both at the same time?
+	};
+
+	// Information about a specific mounted weapon (serialized directly)
+	struct WeaponState {
+		WeaponMount *mount;					// Mounted hardpoint
+		uint8_t group = 0;					// Group this hardpoint belongs to
+		bool withinGimbalLimit = false;		// does the gun have a valid firing solution on the target?
+
+		uint8_t lastBarrel = 0;	 // the last barrel used to fire (for multi-barrel weapons)
+		float temperature = 0;	 // current gun temperature
+		double nextFireTime = 0; // time at which the gun will be ready to fire again
+		// TODO: integrate this with a whole-ship cooling system
+		// Currently uses GunManager::m_coolingBoost
+		// float coolerOverclock = 1.0; // any boost to gun coolers from ship equipment
+		// float powerOverclock = 1.0; // boost to overall firerate (and corresponding increase in heat generation)
+
+		vector3f currentLead;
+		vector3d currentLeadPos;
+
+		WeaponData data;
+		SceneGraph::Model *model; // gun model, currently unused
+	};
+
+	struct WeaponMount {
+		StringName id;
+		SceneGraph::Tag *tag;     // Tag in the parent model that this weapon mount is attached to
+		vector2f gimbalLimitTan;  // tangent of gimbal limits
+		//SceneGraph::Model *model; // model to render for this weapon mount, if any
+		// TODO: enable/disable hardpoint based on ship configuration, i.e. landing/vtol/wings?
+	};
+
+	struct GroupState {
+		WeaponIndexSet weapons;				// Whic weapons are assigned to this group?
+		const Body *target;					// The target for this group, if any
+		bool firing;						// Is the group currently firing
+		bool fireWithoutTargeting;			// Can the group fire without a target/target not in gimbal range?
+		ConnectionTicket onTargetDestroyed;	// Unlock the target once it's been destroyed
+	};
+
+	GunManager() = default;
+
+	void Init(ModelBody *b);
+
+	void SaveToJson(Json &jsonObj, Space *space);
+	void LoadFromJson(const Json &jsonObj, Space *space);
+
+	void StaticUpdate(float deltaTime);
+
+	// ==========================================
+
+	// Add a weapon mount to this gun manager
+	void AddWeaponMount(StringName id, StringName tagName, vector2f gimbalLimitDegrees);
+	// Remove a weapon mount from this gun manager. This will fail if a weapon is still mounted.
+	void RemoveWeaponMount(StringName id);
```

Comment:
How can the caller know if this failed? Should it return a `bool`?

Reply:
The caller should as a matter of convention always validate that the mount is empty (via `IsWeaponMounted`) before calling `RemoveWeaponMount` - as a consequence, I don't think returning an explicit indication of failure here is necessary. This comment is meant to indicate the above precondition rather than indicate that the function is meant to be called in an "exploratory" manner.

Reply:
Ok, so this API is somewhat inconcsistent. Why does MountWeapon() return a bool instead of not also requiring a prior call to IsWeaponMounted()?

And if these really are preconditions and a convention which must be followed, they should be documented, and the convention enforced by asserting the preconditions to prevent misuse by other developers instead of just silently failing.

Reply:
I'll put a boolean return value on this then, as it is a bit inconsistent. The convention is documented in the (updated) function comment, and the only reason the function returns rather than asserting is because a failure to follow the conventions is a recoverable error and asserts are strongly non-recoverable.

Reply:
It becomes a question of whether you want to enforce a convention or not.  If you do, then an assert is the preferred approach as that forces code to adopt the convention instead of relying on reviewers to discover any transgressions. If you don't, then having the capability for client code to gracefully recover is essential. 

And, playing the devil's advocate, if these 2 functions return bool, then they all should if they can fail to complete their intended action.

The worst (IMHO) are silent failures; they'll take forever to discover and debug when some player reports that "something" isn't working.

Reply:
> The worst (IMHO) are silent failures; they'll take forever to discover and debug when some player reports that "something" isn't working.

Agreed. I've modified the return value and added some extra validation in `Ship.lua` when adding weapon mounts.